File: TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/core/module/PicassoModule.java
package com.skedgo.tripkit.ui.core.module;

import android.content.Context;
import com.squareup.picasso.OkHttp3Downloader;
import com.squareup.picasso.Picasso;
import dagger.Module;
import dagger.Provides;
import okhttp3.Cache;
import okhttp3.OkHttpClient;

import javax.inject.Singleton;
import java.io.File;

/**
 * Defines components fetching and loading images.
 */
@Module
public class PicassoModule {
  static Cache createCache(Context context) {
    final File imagesCacheDir = new File(context.getCacheDir(), "picasso-images");
    final int cacheSize = 10 * 1024 * 1024; // 10 MB.
    return new Cache(imagesCacheDir, cacheSize);
  }

  @Provides @Singleton Picasso picasso(
      Context context,
      OkHttpClient httpClient
  ) {
    // Use an own HttpClient in order not to
    // interfere w/ other sorts of requests.
    OkHttpClient.Builder builder = httpClient.newBuilder();
    builder.interceptors().clear();
    builder.networkInterceptors().clear();
    final OkHttpClient downloader = builder
        .cache(createCache(context))
        .build();
    return new Picasso.Builder(context)
        .downloader(new OkHttp3Downloader(downloader))
        .build();
  }
}

File: TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/core/module/PicassoModule.kt
package com.skedgo.tripkit.ui.core.module

import android.content.Context
import com.squareup.picasso.OkHttp3Downloader
import com.squareup.picasso.Picasso
import com.squareup.picasso.Picasso.Builder
import dagger.Module
import dagger.Provides
import okhttp3.Cache
import okhttp3.OkHttpClient
import java.io.File
import javax.inject.Singleton

/**
 * Defines components fetching and loading images.
 */
@Module
class PicassoModule {
    @Provides
    @Singleton
    fun picasso(
        context: Context,
        httpClient: OkHttpClient
    ): Picasso {
        // Use an own HttpClient in order not to
        // interfere w/ other sorts of requests.
        val builder: OkHttpClient.Builder = httpClient.newBuilder()
        builder.interceptors().clear()
        builder.networkInterceptors().clear()
        val downloader: OkHttpClient = builder
            .cache(createCache(context))
            .build()
        return Builder(context)
            .downloader(OkHttp3Downloader(downloader))
            .build()
    }

    companion object {
        fun createCache(context: Context): Cache {
            val imagesCacheDir = File(context.cacheDir, "picasso-images")
            val cacheSize = 10 * 1024 * 1024 // 10 MB.
            return Cache(imagesCacheDir, cacheSize.toLong())
        }
    }
}